Implementation of dissipation balanced convective penetration

mathren opened the pull request:
From [Johnston et al. 2024 ](https://ui.adsabs.harvard.edu/abs/2024ApJ...964..170J/abstract) informed by the 3D simulation by [Anders et al. 2022](https://ui.adsabs.harvard.edu/abs/2022ApJ...926..169A/abstract).

I tested that my re-implementation gives qualitative and quantitative agreement with the original calculations of an 8Msun, solar Z, main sequence star. The plot below shows the mass in the convective penetration zone as a function of time, with in blue the original implementation and inlist (labeled "baseline") , and in orange the output from the example folder in `dbconvpen` (labeled "example rei-mplementation". The two curves are not expected to be numerically identical since I did try to remove some things from the example inlist to simplify, but they are expected to be physically equivalent, and indeed appear:

![image](https://github.com/MESAHub/mesa-contrib/assets/49067756/bc806872-e8a8-46e0-90c7-e9794cad10c4)

 Implementing this in `mesa-contrib` was pre-approved by @colej. Worth tagging also @evananders, @MichielsenM in case they want to chime in.

mathren committed:
Implementation of dissipation balanced convective penetration

From Johnston et al. 2024

warrickball commented:
This looks good but I'm afraid I probably won't be able to test it until at least the end of July. I hope that's not too long to wait!

The only thing I did notice was a couple of old-fashioned comparison operators (e.g. `.lt.` and `.le.`) even though modern style comparisons (e.g. `>=`) are used elsewhere in the code.

mathren committed:
fix old-fashioned comparison operators

plus some more lower-casing

mathren commented:
That is fixed now! And no worries, there is no rush at all

warrickball commented on hooks/dbconvpen/dbconvpen.inc:
```diff
@@ -0,0 +1,341 @@
+    call star_eval_conv_bdy_r(s, 1, r_cb, ierr)
+    s% ixtra(2) = k
+    s% xtra(5) = r_cb
+    s%xtra(1) = s%m(k)
```
```suggestion
    s% xtra(1) = s%m(k)
```

mathren committed:
Update hooks/dbconvpen/dbconvpen.inc

Co-authored-by: Warrick Ball <warrickball@gmail.com>

warrickball commented:
There's a bit of re-arranging I can do once this is merged but I'm satisfied with the overall quality. Some cleanup I'll do later is

* rename `example` to `star_example` (so that it'll work with `check_examples.sh`),
* remove files in `example` that are already part of the work folder, and
* maybe try to reduce the example run time (took me  on 4 cores).

Out of curiosity, do you now have permission to merge this yourself?

mathren committed:
rename example and remove redundant files

mathren committed:
cut by 2x number of timesteps in test case

mathren committed:
add main inlist and cut run time to 4m on single core

main inlist pointing to inlist_dbcp allows for compatibility with
check_examples.sh.

To cut the run time I have removed several options
used in the original paper, these are left as comments
for people to see. Test has been run on my laptop.
